crates/wiggle/generate/src/funcs.rs

```diff
@@ -73,36 +73,55 @@ fn _define_func(
         },
     );
 
-    let asyncness = if settings.get_async(&module, &func).is_sync() {
-        quote!()
-    } else {
-        quote!(async)
-    };
     let mod_name = &module.name.as_str();
     let func_name = &func.name.as_str();
-    (
-        quote! {
+    if settings.get_async(&module, &func).is_sync() {
+        (
+            quote!(
+                #[allow(unreachable_code)] // deals with warnings in noreturn functions
+                pub fn #ident(
+                    ctx: &mut (impl #(#bounds)+*),
+                    memory: &dyn #rt::GuestMemory,
+                    #(#abi_params),*
+                ) -> Result<#abi_ret, #rt::Trap> {
+                    use std::convert::TryFrom as _;
+                    let _span = #rt::tracing::span!(
+                        #rt::tracing::Level::TRACE,
+                        "wiggle abi",
+                        module = #mod_name,
+                        function = #func_name
+                    );
+                    _span.in_scope(|| {
+                      #body
+                    })
+              }
+            ),
+            bounds,
+        )
+    } else {
+        (
+            quote!(
             #[allow(unreachable_code)] // deals with warnings in noreturn functions
-            pub #asyncness fn #ident(
-                ctx: &mut (impl #(#bounds)+*),
-                memory: &dyn #rt::GuestMemory,
+            pub fn #ident<'a>(
+                ctx: &'a mut (impl #(#bounds)+*),
+                memory: &'a dyn #rt::GuestMemory,
                 #(#abi_params),*
-            ) -> Result<#abi_ret, #rt::Trap> {
+            ) -> impl std::future::Future<Output = Result<#abi_ret, #rt::Trap>> + 'a {
                 use std::convert::TryFrom as _;
-
+                use #rt::tracing::Instrument as _;
                 let _span = #rt::tracing::span!(
                     #rt::tracing::Level::TRACE,
                     "wiggle abi",
                     module = #mod_name,
                     function = #func_name
                 );
```

Comment:
Since this call is somewhat nontrivial could this call to `span!` get pulled out of the two branches of this `if` and something like `let _span = #mk_span;` be used in both macros?

Reply:
Yeah! I can do that 👍🏻